ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -850,22 +850,114 @@ int ompi_osc_find_attached_region_position(ompi_osc_dynamic_win_info_t *dynamic_
                                                       base, len, insert);
     }
 }
+inline bool ompi_osc_need_acc_lock(ompi_osc_ucx_module_t *module, int target)
+{
+    ompi_osc_ucx_lock_t *lock = NULL;
+    opal_hash_table_get_value_uint32(&module->outstanding_locks,
+                                     (uint32_t) target, (void **) &lock);
+
+    /* if there is an exclusive lock there is no need to acqurie the accumulate lock */
+    return !(NULL != lock && LOCK_EXCLUSIVE == lock->type);
+}
+
+inline int ompi_osc_state_lock(
+    ompi_osc_ucx_module_t *module,
+    int                    target,
+    bool                  *lock_acquired,
+    bool                   force_lock) {
+    uint64_t result_value = -1;
+    uint64_t remote_addr = (module->state_addrs)[target] + OSC_UCX_STATE_ACC_LOCK_OFFSET;
+    int ret = OMPI_SUCCESS;
+
+    if (force_lock || ompi_osc_need_acc_lock(module, target)) {
+        for (;;) {
+            ret = opal_common_ucx_wpmem_cmpswp(module->state_mem,
+                                            TARGET_LOCK_UNLOCKED, TARGET_LOCK_EXCLUSIVE,
+                                            target, &result_value, sizeof(result_value),
+                                            remote_addr);
+            if (ret != OMPI_SUCCESS) {
+                OSC_UCX_VERBOSE(1, "opal_common_ucx_mem_cmpswp failed: %d", ret);
+                return OMPI_ERROR;
+            }
+            if (result_value == TARGET_LOCK_UNLOCKED) {
+                break;
+            }
+
+            opal_common_ucx_wpool_progress(mca_osc_ucx_component.wpool);
+        }
+
+        *lock_acquired = true;
+    } else {
+        *lock_acquired = false;
+    }
+
+    return OMPI_SUCCESS;
+}
+
+inline int ompi_osc_state_unlock(
```

Comment:
I realize this too late: this should be called `ompi_osc_ucx_state_unlock`. I know this has gone into `main` already, so could you fix that  in a separate PR?

Reply:
@janjust ^ fyi

Reply:
@devreal will do.